### PR TITLE
Introduce samples operator control

### PR DESCRIPTION
**Pull request by petrkotas**
### What this PR does / why we need it:

Add option to control samples operator managed state.
When `registry.redhat.io` key is not present in pull-seceret
and feature gate ManageSamplesOperator in cluster object
is enabled the managed state is flipped to unmanaged.

Signed-off-by: Petr Kotas <pkotas@redhat.com>

### Which issue this PR addresses:

Fixes https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/8749451/

### CODE REVIEW

File: pkg/operator/controllers/pullsecret/pullsecret_controller.go [hunk 1]
```diff
@@ -93,6 +97,32 @@ func (r *PullSecretReconciler) Reconcile(request ctrl.Request) (ctrl.Result, err
...
+	}
+
+	var updated, enabled bool
+	// reconcile the samples operator state
+	// if `registry.redhat.io` pull secret key is missing set samples operator to unmanaged
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		updated, enabled, err = r.updateSamplesController(ctx, redHatKeyCondition)
+		return err
+	})
+
```
Comment (gvanderpotte) [inline pkg/operator/controllers/pullsecret/pullsecret_controller.go:120]: Should not we test for err object here as well ?

**Review by gvanderpotte — COMMENTED** [review]
One small comment, but lgtm so far

File: pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go [hunk 2]
```diff
@@ -57,7 +58,8 @@ type ClusterSpec struct {
 
 // FeaturesSpec defines ARO operator feature gates
 type FeaturesSpec struct {
-	PersistentPrometheus bool `json:"persistentPrometheus,omitempty"`
+	PersistentPrometheus  bool `json:"persistentPrometheus,omitempty"`
+	ManageSamplesOperator bool `json:"manageSamplesOperator,omitempty"`
```
Comment (cadenmarchese) [inline pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go:62]: When this is rebased, we can use this format for the feature flags:
```suggestion
	ReconcileSamplesOperator bool `json:"reconcileSamplesOperator,omitempty"`
```
https://github.com/Azure/ARO-RP/blob/master/pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go#L65-L75

File: pkg/operator/controllers/pullsecret/pullsecret_controller.go [hunk 3]
```diff
@@ -45,14 +47,16 @@ var rhKeys = []string{"registry.redhat.io", "cloud.redhat.com"}
 type PullSecretReconciler struct {
```
Comment (cadenmarchese) [inline pkg/operator/controllers/pullsecret/pullsecret_controller.go:47]:
```suggestion
type Reconciler struct {
```

File: pkg/operator/controllers/pullsecret/pullsecret_controller.go [hunk 4]
```diff
@@ -45,14 +47,16 @@ var rhKeys = []string{"registry.redhat.io", "cloud.redhat.com"}
...
 	kubernetescli kubernetes.Interface
 	arocli        aroclient.Interface
+	samplescli    samplesclient.Interface
 	log           *logrus.Entry
 }
 
-func NewReconciler(log *logrus.Entry, kubernetescli kubernetes.Interface, arocli aroclient.Interface) *PullSecretReconciler {
+func NewReconciler(log *logrus.Entry, kubernetescli kubernetes.Interface, arocli aroclient.Interface, samplescli samplesclient.Interface) *PullSecretReconciler {
 	return &PullSecretReconciler{
```
Comment (cadenmarchese) [inline pkg/operator/controllers/pullsecret/pullsecret_controller.go:55]:
```suggestion
	return &Reconciler{
```

File: pkg/operator/controllers/pullsecret/pullsecret_controller.go [hunk 5]
```diff
@@ -45,14 +47,16 @@ var rhKeys = []string{"registry.redhat.io", "cloud.redhat.com"}
...
 		log:           log,
 		kubernetescli: kubernetescli,
 		arocli:        arocli,
+		samplescli:    samplescli,
```
Comment (cadenmarchese) [inline pkg/operator/controllers/pullsecret/pullsecret_controller.go:59]:
```suggestion
		log:           log,
		
		arocli:        arocli,
		kubernetescli: kubernetescli,
		samplescli:    samplescli,
```

File: pkg/operator/controllers/pullsecret/pullsecret_controller.go [hunk 6]
```diff
@@ -257,3 +287,70 @@ func (r *PullSecretReconciler) keyCondition(failed bool, foundKeys []string) *st
 
 	return keyCondition
 }
+
+func (r *PullSecretReconciler) updateSamplesController(ctx context.Context, state *status.Condition) (updated bool, enabled bool, err error) {
```
Comment (cadenmarchese) [inline pkg/operator/controllers/pullsecret/pullsecret_controller.go:291]:
```suggestion
func (r *Reconciler) updateSamplesController(ctx context.Context, state *status.Condition) (updated bool, enabled bool, err error) {
```

File: pkg/operator/controllers/pullsecret/pullsecret_controller.go [hunk 7]
```diff
@@ -257,3 +287,70 @@ func (r *PullSecretReconciler) keyCondition(failed bool, foundKeys []string) *st
...
+			updated = true
+		}
+		return err
+	})
+
+	return updated, enabled, err
+}
+
+// buildSamplesCondition indicates whether aroOperator modified state of the cluster-samples-operator
+func (r *PullSecretReconciler) buildSamplesCondition(updated bool, foundKey bool) *status.Condition {
```
Comment (cadenmarchese) [inline pkg/operator/controllers/pullsecret/pullsecret_controller.go:330]:
```suggestion
func (r *Reconciler) buildSamplesCondition(updated bool, foundKey bool) *status.Condition {
```

**Review by cadenmarchese — COMMENTED** [review]
I'm recommending changes that were put in place https://github.com/Azure/ARO-RP/pull/1602 and https://github.com/Azure/ARO-RP/pull/1579 be applied to this PR.

**Comment by mjudeikis** [comment]
@petrkotas  what do you want to do with this one?

**Comment by petrkotas** [comment]
Not needed anymore.

**Pull request closed without merging**